### PR TITLE
Clean up subsystem Initialize(), require an explicit result returned, give a formal way to fail (for SSlua)

### PR DESCRIPTION
**Pull request by Tastyfish**
## About The Pull Request

After #69533, the custom error message that SSlua used (eg. if auxtools.dll is missing) was completely broken (just simply showing the number of seconds since midnight) as `start_timeofday` is no longer passed to `subsystem/Initialize()`.

I then realized that literally nothing uses that `start_timeofday` argument except for SSlua failing, period, so I completely cut it out from all of the implementations of SS/Initialize().

I also didn't like that SSlua was just doing its own thing for error handling in the first place, so now there's an explicit result returned for `...subsystem/Initialize()`, similar to `atom/Initialize()`. You no longer need to call the base ..() in subsystem/Initialize() unless you've created an intermediate type that has init code.

Further, if you don't implement Initialize() anymore, you must set the `SS_NO_INIT` flag or your subsystem will throw a warning on init.

The logic is as follows and is derived from MSO's request in the comments below:

Return:
- `SS_INIT_SUCCESS`: Subsystem initialization completes as before.
- `SS_INIT_NO_NEED`: Initialization completes successfully, but nothing is printed to world chat, nor the log.
  - Used if your subsystem detects that it's been disabled or not sufficiently configured to be considered enabled/active to the users, such as with the discord subsystem if not configured.
- `SS_INIT_FAILURE`: The subsystem has canonically failed to initialize and should be suppressed.
   - Subsystem as having not initialized, and firing is disabled. Failure is reflected in the log and world chat.
- `SS_INIT_NONE`: Never intentionally return this, used to detect that Initialize() wasn't actually implemented.
- Any other value: Treated as a bug, or that an unhandled runtime error occurred.
   - For compatibility, the subsystem is treated as fully initialized, but a warning is printed to the console, and the world chat indicates that it initialized with errors.

As a result of implementing this, nearly all subsystems were touched.
- Some were found to previously call ..() at the beginning of their init rather than their end, meaning they were lying about their time metrics (locked to 0 seconds).
   - I only found one instance of this effecting behavior, (spatial grid) so initialized is manually set to TRUE before the rest of the code is written.
- 1 or 2 subsystems didn't previously call ..() at all, evading metrics and logging.
- Achievement and Discord had an early return for lock of configuration, so `SS_INIT_NO_NEED` is appropriately returned.
- Stat Panel and Circuit Components didn't implement Initialize() and don't seem to rely on the value of `initialized` so they are now flagged to `SS_NO_INIT`. They appear to work fine.
   - Conversely, chat code *does* rely on SSchat's `initialized` being set, even though I think it could use the runlevel instead. Testing chat code locally is a nightmare since the relevent code is for handling startup lag, so I'm retaining the old behavior, by adding an initialized that just returns `SS_INIT_SUCCESSS`.
- Mapping and Wiremod Composite previously overrode `New()` instead of the correct `PreInit()`, so that's been corrected.

Because of the sheer number of changes, I'd recommend test merging it.

## Why It's Good For The Game

Gets rid of dead argument in a major proc, SSlua is slightly less rogue.

## Changelog

Error message is technically user/admin-facing, but only if the deployment is broken, so eh.

### CODE REVIEW

File: code/controllers/subsystem.dm [hunk 1]
```diff
@@ -254,17 +254,28 @@
...
+ * failed - Optional: If true, tell everyone that the subsystem failed to init and cancel firing.
+ */
+/datum/controller/subsystem/Initialize(failed = FALSE)
 	initialized = TRUE
-	SEND_SIGNAL(src, COMSIG_SUBSYSTEM_POST_INITIALIZE)
+
+	if(!failed)
+		SEND_SIGNAL(src, COMSIG_SUBSYSTEM_POST_INITIALIZE)
+	else
+		can_fire = FALSE
```
Comment (LemonInTheDark) [inline code/controllers/subsystem.dm:267]: is it really worth making failing to init standard behavior? idk

Comment (Tastyfish) [reply code/controllers/subsystem.dm:267]: I don't know. I just don't like how SSlua handled it by not calling ..() and implementing its own version (which fell out of sync)

Comment (Tastyfish) [reply code/controllers/subsystem.dm:267]: I suppose a reasonable compromise is having SSlua complete normally but also set `can_fire` to false and maybe print an extra message along with the exception. Would you prefer that, @LemonInTheDark ?

Comment (LemonInTheDark) [reply code/controllers/subsystem.dm:267]: let's wait for mso to weigh in, it's his work after all (assuming he isn't high)

Comment (MrStonedOne) [reply code/controllers/subsystem.dm:267]: we do need a way to detect things that don't initialize correctly but given how subsystems are often used as a namespace for global vars, its entirely possible for init and fire to never touch the same variables meaning a failure in one doesn't always begate a failure in others.

I'm thinking enforced return values.

base subsystem init returns -1 (SS_INIT_NONE) (which should print a warning message like the base fire() does)

subsystems should return 2 (SS_INIT_SUCCESS) (prints a world message) or 3 (SS_INIT_NO_NEED) (prints no world message, for if the subsystem is disabled by config or some other condition wasn't met so it did not need to initialize) or -2 (SS_INIT_FAILURE).

returning `-2` should do the can_fire = 0 stuff and print a log message and set the subsystem's initialized to `FALSE`

returning null should be treated as a signal that a runtime happened in the subsystem's init proc and should print a warning message but proceed as if the init was successful.

return anything else should be treated as null with the addition of printing an extra message about invalid init return value.

either way I don't like (ab)using parent calls for signaling like that when we have an outer layer to do signalling properly (with return values).

Comment (Tastyfish) [reply code/controllers/subsystem.dm:267]: I can work on that.

Comment (Tastyfish) [reply code/controllers/subsystem.dm]: @MrStonedOne I've implemented what I believe is the spec you've requested. Due to touching every subsystem, I'd recommend a test merge.

File: code/controllers/subsystem/processing/ai_behaviors.dm [hunk 2]
```diff
@@ -9,9 +9,9 @@ PROCESSING_SUBSYSTEM_DEF(ai_behaviors)
 	///List of all ai_behavior singletons, key is the typepath while assigned value is a newly created instance of the typepath. See SetupAIBehaviors()
 	var/list/ai_behaviors
 
-/datum/controller/subsystem/processing/ai_behaviors/Initialize(timeofday)
+/datum/controller/subsystem/processing/ai_behaviors/Initialize()
 	SetupAIBehaviors()
-	return ..()
+	return SS_INIT_SUCCESS
```
Comment (MrStonedOne) [inline code/controllers/subsystem/processing/ai_behaviors.dm:14]: does `/processing/` have an initialize?

Comment (Tastyfish) [reply code/controllers/subsystem/processing/ai_behaviors.dm:14]: No sir, in fact it's SS_NO_INIT

Comment (Tastyfish) [reply code/controllers/subsystem/processing/ai_behaviors.dm:14]: Same with timer, movement, and verb_manager. Fluids does init but the subtypes don't override.

**Review by MrsTonedOne — APPROVED** [review]

**Comment by Tastyfish** [comment]
fixed merge conflict from #69707

**Comment by MrStonedOne** [comment]
Test merged on sybil with no apparent issues shown in logs or known to me otherwise.

**Review by MrsTonedOne — APPROVED** [review]

**Comment by Tastyfish** [comment]
> Test merged on sybil with no apparent issues shown in logs or known to me otherwise.

Awesome. I tested locally but loads of these subsystems rely on being hit with traffic.

**Comment by JohnFulpWillard** [comment]
> Test merged on sybil with no apparent issues shown in logs or known to me otherwise.

bababooey

**Pull request merged**